tomb-common/src/banyan_api/blockstore.rs

```diff
@@ -1,46 +1,74 @@
 use crate::banyan_api::client::Client;
-use crate::banyan_api::requests::staging::blocks::pull::*;
+use crate::banyan_api::requests::staging::blocks::*;
 use crate::blockstore::BlockStore;
 use anyhow::{anyhow, Result};
 use async_trait::async_trait;
 use futures_util::StreamExt;
-use std::borrow::Cow;
-use std::cell::RefCell;
+use reqwest::Url;
+use std::{
+    borrow::Cow,
+    cell::RefCell,
+    collections::{BTreeSet, HashMap},
+};
 use wnfs::libipld::{Cid, IpldCodec};
 
+use super::requests::core::blocks::locate::LocationRequest;
+
 /// A network-based BlockStore designed to interface with a Kubo node or an API which mirrors it
 
 #[derive(Debug, Clone)]
-pub struct BanyanApiBlockStore(RefCell<Client>);
+pub struct BanyanApiBlockStore {
+    client: Client,
+    /// Known remote endpoints of Blocks
+    block_locations: RefCell<HashMap<String, Vec<String>>>,
+}
 
 impl From<Client> for BanyanApiBlockStore {
     fn from(client: Client) -> Self {
-        Self(RefCell::new(client))
+        Self {
+            client,
+            block_locations: RefCell::new(HashMap::new()),
+        }
+    }
+}
+
+impl BanyanApiBlockStore {
+    /// Find the locations associated with a set of CIDs for fast querying on lookup
+    pub async fn find_cids(&self, cids: BTreeSet<Cid>) -> Result<()> {
+        let request = LocationRequest { cids };
+        let mut client = self.client.clone();
+        let response = client.call(request).await?;
+        let mut block_locations = self.block_locations.borrow_mut();
+        block_locations.extend(response);
+        Ok(())
     }
 }
 
 #[async_trait(?Send)]
 impl BlockStore for BanyanApiBlockStore {
     /// Stores an array of bytes in the block store.
     async fn put_block(&self, _bytes: Vec<u8>, _codec: IpldCodec) -> Result<Cid> {
-        // TODO
-        panic!("Not yet implemented");
+        Err(anyhow!("Cannot put block in API store"))
     }
 
     /// Retrieves an array of bytes from the block store with given CID.
     #[allow(clippy::await_holding_refcell_ref)]
     async fn get_block(&self, cid: &Cid) -> Result<Cow<'_, Vec<u8>>> {
-        let mut client = self.0.borrow_mut();
+        let mut client = self.client.clone();
+        // If there is already a known block location before we do this
+        let base_url = match self.block_locations.borrow().clone().get(&cid.to_string()) {
+            Some(addresses) => Url::parse(&addresses[0])?,
+            None => client.remote_data.clone(),
+        };
```

Comment:
my hopes are that this will lay the ground work to speed up reads in the future. we dont want to have to request a list of locations every time we are unable to find a block. not necessary atm but might be soon

Reply:
Looks good, I agree that it should be done this way